File: python/camera_ui_sdk/sensor/light.py
from __future__ import annotations

from collections.abc import Mapping
from enum import Enum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class LightCapability(str, Enum):
    """Optional capabilities for light controls."""

    Brightness = "brightness"


class LightProperty(str, Enum):
    """Properties for light controls."""

    On = "on"
    Brightness = "brightness"


class LightControlProperties(TypedDict):
    """Property value map for light controls."""

    on: bool
    brightness: int


class LightPropertyChangeData(TypedDict):
    """Emitted on LightControlLike.onPropertyChanged."""

    property: str  # LightProperty value
    value: bool | int


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class LightControlLike(SensorLike, Protocol):
    """Read-only proxy interface for a light control."""

    @property
    def type(self) -> SensorType:
        return SensorType.Light

    @overload
    def getValue(self, property: Literal[LightProperty.On]) -> bool | None: ...
    @overload
    def getValue(self, property: Literal[LightProperty.Brightness]) -> int | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...

    @property
    def onPropertyChanged(self) -> Observable[LightPropertyChangeData]: ...

    @property
    def onCapabilitiesChanged(self) -> Observable[list[LightCapability]]: ...


class LightControl(Sensor[LightControlProperties, TStorage, LightCapability], Generic[TStorage]):
    """Light control sensor. Override ``setOn()``/``setOff()`` to drive your
    hardware, then ``await super().setOn()`` / ``await super().setOff()`` to
    sync the SDK state.

    Plugins that have no hardware-action use case can leave the methods
    unoverridden — the base implementation just updates the state.

    For hardware-pushed updates (someone manually flipped the switch), call
    ``super().setOn()`` / ``super().setOff()`` from your event handler — that
    bypasses any plugin override and only syncs state.
    """

    _requires_frames = False

    def __init__(self, name: str = "Light") -> None:
        super().__init__(name)
        self._write_state(
            {
                LightProperty.On.value: False,
                LightProperty.Brightness.value: 100,
            }
        )

    @property
    def type(self) -> SensorType:
        return SensorType.Light

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Control

    @property
    def on(self) -> bool:
        return bool(self.props.on)

    @property
    def brightness(self) -> int:
        value = self.props.brightness
        return int(value) if value is not None else 0

    async def setOn(self) -> None:
        """Turn the light on. Override to drive hardware and call
        ``await super().setOn()`` after the hardware call succeeds to sync the
        SDK state.

        Example:
            ```python
            await light.setOn()
            ```
        """
        self._write_state({LightProperty.On.value: True})

    async def setOff(self) -> None:
        """Turn the light off. Override to drive hardware and call
        ``await super().setOff()`` after the hardware call succeeds to sync the
        SDK state.

        Example:
            ```python
            await light.setOff()
            ```
        """
        self._write_state({LightProperty.On.value: False})

    async def setBrightness(self, value: int) -> None:
        """Set brightness. Override to drive hardware and call
        ``await super().setBrightness(value)`` after the hardware call
        succeeds. The default implementation clamps the value to [0, 100].

        Args:
            value: Brightness level in the range 0–100.

        Example:
            ```python
            await light.setBrightness(75)
            ```
        """
        self._write_state({LightProperty.Brightness.value: max(0, min(100, value))})

    async def updateValue(self, property: str, value: Any) -> None:
        """Routes generic property writes to semantic methods."""
        if property == LightProperty.On.value:
            if value:
                await self.setOn()
            else:
                await self.setOff()
            return
        if property == LightProperty.Brightness.value:
            await self.setBrightness(int(value))
            return
        # Unknown / non-writable property — ignored.

File: node/src/sensor/light.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/** Optional capabilities of a light control. */
export enum LightCapability {
  /** Light supports brightness adjustment (0-100). */
  Brightness = 'brightness',
}

/**
 * Property names of a light control.
 *
 * @internal
 */
export enum LightProperty {
  /** Whether the light is on. */
  On = 'on',
  /** Brightness level (0-100). */
  Brightness = 'brightness',
}

/**
 * Property values of a light control.
 *
 * @internal
 */
export interface LightControlProperties {
  [LightProperty.On]: boolean;
  [LightProperty.Brightness]: number;
}

/** Read-only proxy interface for a light control. */
export interface LightControlLike extends SensorLike {
  readonly type: SensorType.Light;
  readonly onPropertyChanged: Observable<PropertyChangeOf<LightControlProperties>>;
  readonly onCapabilitiesChanged: Observable<LightCapability[]>;

  getValue(property: LightProperty.On): boolean | undefined;
  getValue(property: LightProperty.Brightness): number | undefined;
  getValue(property: string): unknown;
}

/**
 * Light control sensor. Override `setOn()`/`setOff()` to drive your hardware,
 * then `await super.setOn()` / `await super.setOff()` to sync the SDK state.
 *
 * Plugins with no hardware-action use case can leave the methods unoverridden,
 * the base implementation just updates the state.
 *
 * For hardware-pushed updates (someone manually flipped the switch), call
 * `super.setOn()` / `super.setOff()` from your event handler. That bypasses any
 * plugin override and only syncs state.
 */
export class LightControl<TStorage extends object = Record<string, any>> extends Sensor<LightControlProperties, TStorage, LightCapability> {
  readonly type = SensorType.Light;
  readonly category = SensorCategory.Control;

  constructor(name = 'Light', options?: SensorOptions) {
    super(name, options);

    this._writeState({
      [LightProperty.On]: false,
      [LightProperty.Brightness]: 100,
    });
  }

  get on(): boolean {
    return this.props.on;
  }

  get brightness(): number {
    return this.props.brightness;
  }

  /**
   * Turn the light on. Override to drive hardware and call `await super.setOn()`
   * after the hardware call succeeds to sync the SDK state.
   *
   * @example
   * ```ts
   * await light.setOn();
   * ```
   */
  async setOn(): Promise<void> {
    this._writeState({ [LightProperty.On]: true });
  }

  /**
   * Turn the light off. Override to drive hardware and call `await super.setOff()`
   * after the hardware call succeeds to sync the SDK state.
   *
   * @example
   * ```ts
   * await light.setOff();
   * ```
   */
  async setOff(): Promise<void> {
    this._writeState({ [LightProperty.On]: false });
  }

  /**
   * Set brightness. Override to drive hardware and call `await super.setBrightness(value)`
   * after the hardware call succeeds. The default implementation clamps the value to [0, 100].
   *
   * @param value - Brightness level in the range 0-100.
   *
   * @example
   * ```ts
   * await light.setBrightness(75);
   * ```
   */
  async setBrightness(value: number): Promise<void> {
    this._writeState({ [LightProperty.Brightness]: Math.max(0, Math.min(100, value)) });
  }

  /**
   * Routes generic property writes to the semantic setters. Only `on` and `brightness` are externally writable.
   *
   * @internal
   */
  async updateValue(property: string, value: unknown): Promise<void> {
    switch (property as LightProperty) {
      case LightProperty.On:
        if (value) await this.setOn();
        else await this.setOff();
        return;
      case LightProperty.Brightness:
        await this.setBrightness(value as number);
        return;
    }
  }
}

/** Registry metadata for {@link LightControl}. */
export const lightMeta = defineSensor({
  type: SensorType.Light,
  category: SensorCategory.Control,
  assignmentKey: 'light',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [LightProperty.On]: { type: 'boolean', writable: true },
    [LightProperty.Brightness]: { type: 'number', min: 0, max: 100, unit: '%', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: LightProperty.On, value: true, sustained: true },
  propertyCapabilities: { [LightProperty.Brightness]: LightCapability.Brightness },
  virtual: {
    properties: { [LightProperty.On]: false, [LightProperty.Brightness]: 100 },
    capabilities: [LightCapability.Brightness],
  },
  semantics: {
    domain: SensorDomain.Light,
    stateProperty: LightProperty.On,
    commandProperty: LightProperty.On,
  },
});
